tools/Azure.Mcp.Tools.VirtualDesktop/src/Services/VirtualDesktopService.cs

```diff
@@ -1,165 +1,342 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using System.Text.Json;
+using System.Threading;
+using Azure.Core;
 using Azure.Mcp.Core.Options;
+using Azure.Mcp.Core.Services.Azure;
 using Azure.Mcp.Core.Services.Azure.Subscription;
+using Azure.Mcp.Core.Services.Azure.Tenant;
 using Azure.Mcp.Tools.VirtualDesktop.Models;
-using Azure.ResourceManager.DesktopVirtualization;
+using Microsoft.Extensions.Logging;
 
 namespace Azure.Mcp.Tools.VirtualDesktop.Services;
 
-public class VirtualDesktopService(ISubscriptionService subscriptionService) : IVirtualDesktopService
+public class VirtualDesktopService(ISubscriptionService subscriptionService, ITenantService tenantService, ILogger<VirtualDesktopService> logger) : BaseAzureResourceService(subscriptionService, tenantService), IVirtualDesktopService
 {
     private readonly ISubscriptionService _subscriptionService = subscriptionService;
+    private readonly ILogger<VirtualDesktopService> _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
     public async Task<IReadOnlyList<HostPool>> ListHostpoolsAsync(string subscription, string? tenant = null, RetryPolicyOptions? retryPolicy = null)
     {
-        var sub = await _subscriptionService.GetSubscription(subscription, tenant, retryPolicy);
-        var hostpools = new List<HostPool>();
-        await foreach (HostPoolResource resource in sub.GetHostPoolsAsync())
+        ValidateRequiredParameters((nameof(subscription), subscription));
+
+        try
+        {
+            return await ExecuteResourceQueryAsync(
+                "Microsoft.DesktopVirtualization/hostPools",
+                null,
+                subscription,
+                retryPolicy,
+                ConvertToHostPoolModel);
+        }
+        catch (Exception ex)
         {
-            hostpools.Add(new HostPool(resource));
+            _logger.LogError(ex, "Error listing Host Pools in Subscription: {Subscription}", subscription);
+            throw;
         }
-        return hostpools;
     }
 
     public async Task<IReadOnlyList<HostPool>> ListHostpoolsByResourceGroupAsync(string subscription, string resourceGroup, string? tenant = null, RetryPolicyOptions? retryPolicy = null)
```

Comment:
Does this need to be a separate method anymore? Couldn't `string resourceGroup` be changed to `string? resourceGroup = null` and then this be merged with the method above.